File: web/frontend/src/api/test-mode.ts
import { launcherFetch } from "@/api/http"

export interface TestModeStatus {
  status: "not_configured" | "in_test" | "production" | string
  in_test: boolean
  completed_at?: string
  completed_by?: string
  completed_source?: string
  active_profile: string
  allow_from: string[]
  can_finish: boolean
  blocked_by: string[]
}

export interface FinishTestModeResponse {
  finished: boolean
  reason?: string
  status: TestModeStatus
}

export async function getTestModeStatus(): Promise<TestModeStatus> {
  const res = await launcherFetch("/api/workspace/test-mode", {
    cache: "no-store",
  })
  if (!res.ok) {
    throw new Error(`API error: ${res.status}`)
  }
  return res.json() as Promise<TestModeStatus>
}

export async function finishTestMode(): Promise<FinishTestModeResponse> {
  const res = await launcherFetch("/api/workspace/test-mode/finish", {
    method: "POST",
  })
  if (!res.ok) {
    const data = (await res.json().catch(() => null)) as
      | FinishTestModeResponse
      | null
    if (data) return data
    throw new Error(`API error: ${res.status}`)
  }
  return res.json() as Promise<FinishTestModeResponse>
}

File: web/backend/api/test_mode.go
package api

import (
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"path/filepath"
	"strings"

	saastenant "github.com/sipeed/picoclaw/internal/saas/tenant"
	"github.com/sipeed/picoclaw/pkg/config"
)

func (h *Handler) registerTestModeRoutes(mux *http.ServeMux) {
	mux.HandleFunc("GET /api/workspace/test-mode", h.handleGetTestMode)
	mux.HandleFunc("POST /api/workspace/test-mode/finish", h.handleFinishTestMode)
}

func (h *Handler) handleGetTestMode(w http.ResponseWriter, _ *http.Request) {
	volumeRoot, err := h.currentTenantVolumeRoot()
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	status, err := saastenant.ReadTestModeStatus(volumeRoot)
	if err != nil {
		http.Error(w, fmt.Sprintf("Failed to load test mode: %v", err), http.StatusInternalServerError)
		return
	}
	writeJSON(w, status)
}

func (h *Handler) handleFinishTestMode(w http.ResponseWriter, _ *http.Request) {
	volumeRoot, err := h.currentTenantVolumeRoot()
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	status, err := saastenant.FinishTestMode(volumeRoot, saastenant.FinishTestModeInput{
		CompletedBy:              "tenant_owner",
		CompletedSource:          "owner",
		RequireWhatsAppAllowlist: true,
	})
	if err != nil {
		if errors.Is(err, saastenant.ErrWhatsAppAllowlistRequired) {
			w.Header().Set("Content-Type", "application/json")
			w.Header().Set("Cache-Control", "no-store")
			w.WriteHeader(http.StatusUnprocessableEntity)
			_ = json.NewEncoder(w).Encode(map[string]any{
				"finished": false,
				"reason":   "whatsapp_allowlist_empty",
				"status":   status,
			})
			return
		}
		http.Error(w, fmt.Sprintf("Failed to finish test mode: %v", err), http.StatusInternalServerError)
		return
	}
	writeJSON(w, map[string]any{
		"finished": true,
		"status":   status,
	})
}

func (h *Handler) currentTenantVolumeRoot() (string, error) {
	cfg, err := config.LoadConfig(h.configPath)
	if err != nil {
		return "", fmt.Errorf("Failed to load config: %w", err)
	}
	workspace := strings.TrimSpace(cfg.WorkspacePath())
	if workspace == "" {
		return filepath.Dir(h.configPath), nil
	}
	if filepath.Base(workspace) == "workspace" {
		return filepath.Dir(workspace), nil
	}
	return filepath.Dir(h.configPath), nil
}
